keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,273 +122,289 @@ consistent with the syntax supported by any of the common proxy vendors:
-and PathPrefix in the API.
-
-<!--
-1) v1beta1 gets a "PathPrefix" field, specced as a strict prefix, no
-trailing "*" or ".*" needed.  Document that Path is under-specified
-and implementation-specific.  Document that all implementations are
-expected to support PathPrefix with consistent semantics.  Path and
-PathPrefix can be mutually exclusive.  v1 gets either Path or
-PathPrefix (debatable) and an annotation
-"ingress.kubernetes.io/beta-path" which maps the new field to the old.
-That satisfies round-trip requirement but is kind of horrible.
-
-2) v1beta1 gets a "PathPrefix" field as above.  v1 gets an annotation
-"ingress.kubernetes.io/beta-path-semantics".  If we create a v1 object
-from a v1beta1 where `path` was used, we set the annotation.
-Validation logic will check the annotation to decide how to interpret
-and validate paths.  We'll have to keep it for a long while.  Maybe we
-can prevent new objects from being created with that annotation?  Now
-what if implementations want to offer "extended" semantics here?  Same
-validation problem.
-
-3)  v1beta1 gets a "PathPrefix" field as above.  v1 gets an annotation
-"ingress.kubernetes.io/non-portable-path-semantics".  If we create a
-v1 object from a v1beta1 where `path` was used, we set the annotation.
-We keep it forever.
-
-4) Same as 2 but we just don't validate Path at all.
--->
-
-#### Rejected alternatives
-
-##### Portable regex behavior
-
-The safest route for specifying the regex would be to state a limited
-subset that can be used in a portable way. Any expressions outside of
-the subset will have implementation specific behavior.
-
-Regular expression subset (derived from [re2][re2-syntax] syntax page)
-
-| Expression | description             |
-|------------|-------------------------|
-| `.`        | any character           |
-| `[xyz]`    | character class         |
-| `[^xyz]`   | negated character class |
-| `x*`       | 0 or more x's           |
-| `x+`       | 1 or more x's           |
-| `xy`       | x followed by y         |
-| `x|y`      | x or y (prefer x)       |
-| `(abc)`    | grouping                |
+  // Defaults to "/" if empty.
+  //
+  // +Optional
+  Path string
 
-Maintaining a regular expression subset is likely not worth the complexity and
-is likely impossible across the [many implementations][regex-survey].
+  // PathType determines the interpretation of the Path
+  // matching. PathType can be one of the following values:
+  //
+  // Exact  - matches the URL path exactly.
+  // Prefix - matches based on a URL path prefix split
+  //   by '/'. [insert description of semantics described below]
+  //
+  // ImplementationSpecific - interpretation of the Path
+  //   matching is up to the IngressClass. Not all
+  //
+  // +Optional
+  PathType string
+  ...
+ }
+ ```
+
+Interoperability between versions will be as follows:
+
+1. v1beta1 will default to `ImplementationSpecific`.
+1. v1 will default to `Prefix`.
```

Comment:
Defaulting to a different value in a new version is perfectly fine. The emphasis on the rule is on "**currently** supported versions", so that the current implicit behavior is maintained.

Reply:
That said, would it be better to make v1 writers specify a PathType (e.g. make it required with no default in v1)? The downside is it makes every consumer be explicit in v1, which makes the manifests more verbose. The upside is it makes every consumer be explicit in v1, which reduces surprise.

Reply:
That is a handy doc -- is there any plan to merge the content into the API conventions doc? The current "defaults" section is pretty short.

Reply:
Changing to require being explicit, let's see how we feel about the (slight) increase in verbosity.

Reply:
Actually -- wouldn't defaulting to ImplementationSpecific be more behavior preserving?